dice_ml/explainer_interfaces/explainer_base.py

```diff
@@ -51,26 +52,87 @@ def do_posthoc_sparsity_enhancement(self, final_cfs_sparse, cfs_preds_sparse, qu
-                    while((abs(diff)>10e-4) and (np.sign(diff*old_diff) > 0) and
-                          ((self.target_cf_class == 0 and current_pred < self.stopping_threshold) |
-                           (self.target_cf_class == 1 and current_pred > self.stopping_threshold))):
-                        old_val = final_cfs_sparse[cf_ix].ravel()[feat_ix]
-                        final_cfs_sparse[cf_ix].ravel()[feat_ix] += np.sign(diff)*change
-                        current_pred = self.predict_fn(final_cfs_sparse[cf_ix])
-                        old_diff = diff
+                    if posthoc_sparsity_algorithm == "linear":
+                        final_cfs_sparse[cf_ix] = do_linear_search(self, diff, decimal_prec, query_instance, cf_ix, feat_ix, final_cfs_sparse, current_pred)
 
-                        if(((self.target_cf_class == 0 and current_pred > self.stopping_threshold) | (self.target_cf_class == 1 and current_pred < self.stopping_threshold))):
-                            final_cfs_sparse[cf_ix].ravel()[feat_ix] = old_val
-                            diff = query_instance.ravel()[feat_ix] - final_cfs_sparse[cf_ix].ravel()[feat_ix]
-                            break
-
-                        diff = query_instance.ravel()[feat_ix] - final_cfs_sparse[cf_ix].ravel()[feat_ix]
+                    elif posthoc_sparsity_algorithm == "binary":
+                        final_cfs_sparse[cf_ix] = do_binary_search(self, diff, decimal_prec, query_instance, cf_ix, feat_ix, final_cfs_sparse, current_pred)
 
             cfs_preds_sparse[cf_ix] = self.predict_fn(final_cfs_sparse[cf_ix])
 
         return final_cfs_sparse, cfs_preds_sparse
+
+def do_linear_search(self, diff, decimal_prec, query_instance, cf_ix, feat_ix, final_cfs_sparse, current_pred):
+
+    old_diff = diff
+    change = (10**-decimal_prec[feat_ix])/(self.cont_maxx[feat_ix] - self.cont_minx[feat_ix])
+    while((abs(diff)>10e-4) and (np.sign(diff*old_diff) > 0) and
+          ((self.target_cf_class == 0 and current_pred < self.stopping_threshold) |
+           (self.target_cf_class == 1 and current_pred > self.stopping_threshold))):
+        old_val = final_cfs_sparse[cf_ix].ravel()[feat_ix]
+        final_cfs_sparse[cf_ix].ravel()[feat_ix] += np.sign(diff)*change
+        current_pred = self.predict_fn(final_cfs_sparse[cf_ix])
+        old_diff = diff
+
+        if(((self.target_cf_class == 0 and current_pred > self.stopping_threshold) | (self.target_cf_class == 1 and current_pred < self.stopping_threshold))):
+            final_cfs_sparse[cf_ix].ravel()[feat_ix] = old_val
+            diff = query_instance.ravel()[feat_ix] - final_cfs_sparse[cf_ix].ravel()[feat_ix]
+            return final_cfs_sparse[cf_ix]
+
+        diff = query_instance.ravel()[feat_ix] - final_cfs_sparse[cf_ix].ravel()[feat_ix]
+
+    return final_cfs_sparse[cf_ix]
+
+def do_binary_search(self, diff, decimal_prec, query_instance, cf_ix, feat_ix, final_cfs_sparse, current_pred):
```

Comment:
some comments in the code will be useful to understand the logic of the algorithm.